streamingccl: support AddSSTable operation in producer processor

gh-casper opened the pull request:
This PR makes producer processor (event stream value generator)
support RangefeedSSTable event, prune it and send it over to
consumers in the ingestion side.

Release note: None

Closes: https://github.com/cockroachdb/cockroach/issues/79309

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/80450)
<!-- Reviewable:end -->

gh-casper reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80450)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @miretskiy)

___
*[`pkg/ccl/streamingccl/streamproducer/replication_stream_test.go` line 376 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N0OzBWn4g9ldcOy2-MB:-N0OzBWn4g9ldcOy2-MC:b-wqe683) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3576ca5412e7d5b67bc63e9715b6d8497e608a9/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go#L376)):*
> ```Go
> 
> 
> // Test results observed:
> ```

This PR is still DRAFT mode. I want to understand some rangefeed behaviors after I encountered some test failures.

This test tries to create rangefeed before or after the import operation (addsstable operation). I should be able to see the sstable on the rangefeed client in both cases? Now, I only see sstable when import is after the rangefeed creation.

In the case I do see sstable, there are some other RangeFeedValue events that carry the same data as well. I'm wondering if there is any misuse of setting when we add sstable in the import code?

https://github.com/cockroachdb/cockroach/blob/e3576ca5412e7d5b67bc63e9715b6d8497e608a9/pkg/kv/bulk/sst_batcher.go#L536-L548


<!-- Sent from Reviewable.io -->

miretskiy commented on pkg/kv/kvserver/rangefeed/registry.go:
```diff
@@ -455,6 +456,16 @@ func (reg *registry) PublishToOverlapping(
 		panic(fmt.Sprintf("unexpected RangeFeedEvent variant: %v", t))
 	}
 
+	if event.SST != nil {
```
this needs to be reverted.

miretskiy commented on pkg/kv/bulk/sst_batcher.go:
```diff
@@ -41,7 +41,7 @@ var (
 		"kv.bulk_io_write.small_write_size",
 		"size below which a 'bulk' write will be performed as a normal write instead",
-		400*1<<10, // 400 Kib
+		1, // 400 Kib
```
looks like it's a test only change. Clearly can't go in

miretskiy commented on pkg/ccl/streamingccl/streamproducer/event_stream.go:
```diff
@@ -340,6 +348,37 @@ func (s *eventStream) streamLoop(ctx context.Context, frontier *span.Frontier) e
 		batchSize += keyValue.Size()
 	}
 
+	scanSST := func(sst *roachpb.RangeFeedSSTable) error {
```
This can be moved to a separate function.

miretskiy commented on pkg/ccl/streamingccl/streamproducer/event_stream.go:
```diff
@@ -340,6 +348,37 @@ func (s *eventStream) streamLoop(ctx context.Context, frontier *span.Frontier) e
+						Timestamp: iter.UnsafeKey().Timestamp,
+					},
+				}
+				batch.KeyValues = append(batch.KeyValues, keyValue)
```
Doesn't it defeat the purpose of ssts -- i.e. exploding them into key/values? That's already happening btw; If you need to trim things, perhaps you should either regenerate sstable... Or... perhaps do it on the receiving end (it's true -- there is more data sent, but the sender usually needs to rekey anyway -- so will be iterating sst anyway.

Maybe there is another way of doing this altogether -- but sending 1 key at a time is not it.

erikgrinaker replied:
+1. The target cluster should ingest these SSTs directly, similarly to the source cluster. Otherwise, if the source is running a large import and the target cluster ingests these as regular point writes, the target cluster won't be able to keep up at all.

erikgrinaker replied:
Which reminds me: we may want to use a similar optimization for initial/catchup scans during streaming replication, such that this data is bulk loaded rather than ingested as point writes.

miretskiy replied:
Hmm.. Yeah, I guess we really should emit initial scan as ssts.  Probably something that should go into rangefeed library.

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80450)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @gh-casper, and @miretskiy)

___
*[`pkg/ccl/streamingccl/streamproducer/replication_stream_test.go` line 376 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N0OzBWn4g9ldcOy2-MB:-N0Qe1JvA15EtEQdjYNN:b-gcdxg5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3576ca5412e7d5b67bc63e9715b6d8497e608a9/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go#L376)):*
> This test tries to create rangefeed before or after the import operation (addsstable operation). I should be able to see the sstable on the rangefeed client in both cases? Now, I only see sstable when import is after the rangefeed creation.

No, it will only be visible when the rangefeed was set up before the `AddSSTable` operations. After the operation is applied, the persisted data is identical to data written via e.g. `Put`, and will be emitted as regular values in an initial/catchup scan.

The `RangeFeedSSTable` event emission is only a performance optimization, which avoids having to decode the SST and emit lots of `RangeFeedValue` events in the Raft hot path.


<!-- Sent from Reviewable.io -->

gh-casper reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80450)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @gh-casper, and @miretskiy)

___
*[`pkg/ccl/streamingccl/streamproducer/replication_stream_test.go` line 376 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N0OzBWn4g9ldcOy2-MB:-N0TOnUk0kcB9_CkxmIW:b9q6bbe) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3576ca5412e7d5b67bc63e9715b6d8497e608a9/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go#L376)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

> This test tries to create rangefeed before or after the import operation (addsstable operation). I should be able to see the sstable on the rangefeed client in both cases? Now, I only see sstable when import is after the rangefeed creation.

No, it will only be visible when the rangefeed was set up before the `AddSSTable` operations. After the operation is applied, the persisted data is identical to data written via e.g. `Put`, and will be emitted as regular values in an initial/catchup scan.

The `RangeFeedSSTable` event emission is only a performance optimization, which avoids having to decode the SST and emit lots of `RangeFeedValue` events in the Raft hot path.
</blockquote></details>

Regarding "In the case I do see sstable, there are some other RangeFeedValue events that carry the same data as well. I'm wondering if there is any misuse of setting when we add sstable in the import code?":

So in rangefeed, if we never emit RangefeedValue events that carry the same value in sst if we emit an sstable at the same time? 


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80450)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @gh-casper, and @miretskiy)

___
*[`pkg/ccl/streamingccl/streamproducer/replication_stream_test.go` line 376 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N0OzBWn4g9ldcOy2-MB:-N0Uwunq-H0SXq0NxRDa:b-p9ymqv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3576ca5412e7d5b67bc63e9715b6d8497e608a9/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go#L376)):*
<details><summary><i>Previously, gh-casper wrote…</i></summary><blockquote>

Regarding "In the case I do see sstable, there are some other RangeFeedValue events that carry the same data as well. I'm wondering if there is any misuse of setting when we add sstable in the import code?":

So in rangefeed, if we never emit RangefeedValue events that carry the same value in sst if we emit an sstable at the same time? 




</blockquote></details>

Correct, if we emit an SST then we don't emit RangefeedValue for any contents. I can think of a couple of other reasons why you saw these duplicates:

* A catchup scan was triggered before a checkpoint, which emitted the same values as RangefeedValue.
* The importer wrote the values again, either as regular put requests or as an SST that was small enough to trigger [`IngestAsWrites`](https://github.com/cockroachdb/cockroach/blob/fa596bc6a4d4cb985405c1947e0ca7349bd95962/pkg/roachpb/api.proto#L1769-L1776).


<!-- Sent from Reviewable.io -->

miretskiy commented on pkg/ccl/streamingccl/streamproducer/event_stream.go:
```diff
@@ -250,6 +253,11 @@ func (s *eventStream) onSpanCompleted(ctx context.Context, sp roachpb.Span) erro
 }
 
+func (s *eventStream) onSSTable(ctx context.Context, sst *roachpb.RangeFeedSSTable) {
+	fmt.Println("eventloop: received a sst")
```
Stray comment?

miretskiy commented on pkg/ccl/streamingccl/streamproducer/event_stream.go:
```diff
@@ -324,6 +332,48 @@ func (p *checkpointPacer) shouldCheckpoint(
+	}
+	defer iter.Close()
+	size := 0
+	for _, sp := range s.spec.Spans {
```
I don't think we guaranteed Spans sorted.. Would it make sense to sort them using SpangGroup first?

miretskiy commented on pkg/ccl/streamingccl/streamproducer/event_stream.go:
```diff
@@ -324,6 +332,48 @@ func (p *checkpointPacer) shouldCheckpoint(
+		endKey := storage.MVCCKey{Key: sp.EndKey}
+		for ; ; iter.Next() {
+			if ok, err := iter.Valid(); err != nil {
+				return size, err
```
`return 0, err` please.

miretskiy commented on pkg/ccl/streamingccl/streamproducer/event_stream.go:
```diff
@@ -324,6 +332,48 @@ func (p *checkpointPacer) shouldCheckpoint(
+				break
+			}
+			keyValue := roachpb.KeyValue{
+				Key: iter.UnsafeKey().Key,
```
I'm pretty sure it is not safe to call UnsafeKey.  I'm also surprised that none of the tests failed.  UnsafeKey returns "unsafe" buffer -- buffer that may be reset on the next call to iter.Next -- yet, you're storing these values inside batch.

miretskiy commented on pkg/ccl/streamingccl/streamproducer/event_stream.go:
```diff
@@ -384,8 +435,17 @@ func (s *eventStream) streamLoop(ctx context.Context, frontier *span.Frontier) e
 					}
 				}
+			case ev.SST != nil:
+				fmt.Println("received a sst in event loop")
```
stray printf

miretskiy commented on pkg/ccl/streamingccl/streamproducer/replication_stream_test.go:
```diff
@@ -367,3 +368,90 @@ CREATE TABLE t2(
+					}
+					require.LessOrEqual(t, 1, len(codec.e.Batch.KeyValues))
+					require.Regexp(t, "/Tenant/10/Table/.*42.*", codec.e.Batch.KeyValues[0].Key.String())
+				}
```
how do you know it's table ID 42?

miretskiy commented on pkg/ccl/streamingccl/streamproducer/replication_stream_test.go:
```diff
@@ -367,3 +368,90 @@ CREATE TABLE t2(
+		for {
+			require.True(t, source.rows.Next())
+			source.codec.decode()
+			if codec.e.Batch != nil {
```
How do you know if addSSt requests were issued? Looks like data server produces just 1 line? that will not result in addsst requests, afaik.

gh-casper committed:
streamingccl: support AddSSTable operation in producer processor

This PR makes producer processor (event stream value generator)
support RangefeedSSTable event, prune it and send it over to
consumers in the ingestion side.

Release note: None

gh-casper reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80450)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @miretskiy)

___
*[`pkg/ccl/streamingccl/streamproducer/event_stream.go` line 257 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N0OyuM1H3qMUL4CGzTm-r1-257:-N2D89rNA7D1TSOU_hGM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3576ca5412e7d5b67bc63e9715b6d8497e608a9/pkg/ccl/streamingccl/streamproducer/event_stream.go#L257)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Stray comment?
</blockquote></details>

Done.
___
*[`pkg/ccl/streamingccl/streamproducer/event_stream.go` line 351 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N0OyuM1H3qMUL4CGzTm-r1-351:-N2D8tik9CAxufVgejpl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3576ca5412e7d5b67bc63e9715b6d8497e608a9/pkg/ccl/streamingccl/streamproducer/event_stream.go#L351)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

This can be moved to a separate function.
</blockquote></details>

Done.
___
*[`pkg/ccl/streamingccl/streamproducer/event_stream.go` line 375 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N0OyuM1H3qMUL4CGzTm-r1-375:-N2DGzUP6f3LRB1SEZMF:b-bji9vf) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3576ca5412e7d5b67bc63e9715b6d8497e608a9/pkg/ccl/streamingccl/streamproducer/event_stream.go#L375)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Hmm.. Yeah, I guess we really should emit initial scan as ssts.  Probably something that should go into rangefeed library.
</blockquote></details>

So right now the approach is that we emit sstables if the spec spans contains the rangefeed sstableevent span. If not, we emit key values to avoid regenerating a new sstable which will be read key by key and rekeyed in downstream anyway.
___
*[`pkg/ccl/streamingccl/streamproducer/event_stream.go` line 352 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N0OyuM1H3qMUL4CGzTm-r2-352:-N2DBbfTBuGCl73iHYX2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6eeaca5de164264f0d951d7ceb4a359f67c0a52/pkg/ccl/streamingccl/streamproducer/event_stream.go#L352)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I don't think we guaranteed Spans sorted.. Would it make sense to sort them using SpangGroup first?
</blockquote></details>

Done.
___
*[`pkg/ccl/streamingccl/streamproducer/event_stream.go` line 357 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N0OyuM1H3qMUL4CGzTm-r2-357:-N2DBgfM6IvUA9kuV3lC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6eeaca5de164264f0d951d7ceb4a359f67c0a52/pkg/ccl/streamingccl/streamproducer/event_stream.go#L357)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

`return 0, err` please.
</blockquote></details>

Done.
___
*[`pkg/ccl/streamingccl/streamproducer/event_stream.go` line 364 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N0OyuM1H3qMUL4CGzTm-r2-364:-N2DGhJT4b1uVcX0N4Y8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6eeaca5de164264f0d951d7ceb4a359f67c0a52/pkg/ccl/streamingccl/streamproducer/event_stream.go#L364)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I'm pretty sure it is not safe to call UnsafeKey.  I'm also surprised that none of the tests failed.  UnsafeKey returns "unsafe" buffer -- buffer that may be reset on the next call to iter.Next -- yet, you're storing these values inside batch.
</blockquote></details>

Done.
___
*[`pkg/ccl/streamingccl/streamproducer/event_stream.go` line 439 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N0OyuM1H3qMUL4CGzTm-r2-439:-N2DD5Q-AsiYnjQG-sUQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6eeaca5de164264f0d951d7ceb4a359f67c0a52/pkg/ccl/streamingccl/streamproducer/event_stream.go#L439)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

stray printf
</blockquote></details>

Done.
___
*[`pkg/kv/bulk/sst_batcher.go` line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N0OyuM1H3qMUL4CGzTo-r1-44:-N2DHcoW3OXfoOj1Jqpf:bwvrgc9) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3576ca5412e7d5b67bc63e9715b6d8497e608a9/pkg/kv/bulk/sst_batcher.go#L44)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

looks like it's a test only change. Clearly can't go in
</blockquote></details>

The original setting was 400 Kib. Wondering why it shows changed on your side.
___
*[`pkg/ccl/streamingccl/streamproducer/replication_stream_test.go` line 422 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N0OyuM1H3qMUL4CGzTn-r2-422:-N2DItt45Ct_QIuKwvGe:b4jxaow) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6eeaca5de164264f0d951d7ceb4a359f67c0a52/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go#L422)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

How do you know if addSSt requests were issued? Looks like data server produces just 1 line? that will not result in addsst requests, afaik.
</blockquote></details>

I previously did an import for one line "42, 42" and set the "kv.bulk_io_write.small_write_size" to be one which makes an import of any size to be an AddSSTable operation.
___
*[`pkg/ccl/streamingccl/streamproducer/replication_stream_test.go` line 432 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N0OyuM1H3qMUL4CGzTn-r2-432:-N2DHznp3fdwd8cV4oXf:blc2jyv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6eeaca5de164264f0d951d7ceb4a359f67c0a52/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go#L432)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

how do you know it's table ID 42?
</blockquote></details>

42 is not the table id, it is the row value I inserted at the beginning.
___
*[`pkg/kv/kvserver/rangefeed/registry.go` line 459 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N0OyuM1H3qMUL4CGzTp-r1-459:-N2DFm5IAMk_T6ZGwV3Z:bpf1bc9) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3576ca5412e7d5b67bc63e9715b6d8497e608a9/pkg/kv/kvserver/rangefeed/registry.go#L459)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

this needs to be reverted.
</blockquote></details>

Why? this code was not touched.



_Code quote:_
```Go
unexpected RangeFeedEvent 
```


<!-- Sent from Reviewable.io -->

miretskiy reviewed (changes requested):
> Reviewed 2 of 11 files at r2, 2 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80450)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @gh-casper, and @miretskiy)

___
*[`pkg/ccl/streamingccl/streamproducer/event_stream.go` line 258 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N2M7FSh5U2yR5a0Npyi:-N2M7FSh5U2yR5a0Npyj:b-o8ge9) ([raw file](https://github.com/cockroachdb/cockroach/blob/3eebe0542439384d29c4e9e2e1e1c2e7482533a9/pkg/ccl/streamingccl/streamproducer/event_stream.go#L258)):*
> ```Go
> 
> func (s *eventStream) onSSTable(ctx context.Context, sst *roachpb.RangeFeedSSTable) {
> 	s.eventsCh <- roachpb.RangeFeedEvent{SST: sst}
> ```

you probably want to select to make sure you detect ctx.Done
___
*[`pkg/ccl/streamingccl/streamproducer/event_stream.go` line 335 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N2M8fCV9h48bu-PfC9D:-N2M8fCV9h48bu-PfC9E:bokltud) ([raw file](https://github.com/cockroachdb/cockroach/blob/3eebe0542439384d29c4e9e2e1e1c2e7482533a9/pkg/ccl/streamingccl/streamproducer/event_stream.go#L335)):*
> ```Go
> }
> 
> func (s *eventStream) trimSST(sst *roachpb.RangeFeedSSTable, batch *streampb.StreamEvent_Batch) (int, error) {
> ```

add a function comment please?
Also wrap long lines if you can.
___
*[`pkg/ccl/streamingccl/streamproducer/event_stream.go` line 365 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N2M8Nre5yv_cegN4YDU:-N2M8Nre5yv_cegN4YDV:b28p2sk) ([raw file](https://github.com/cockroachdb/cockroach/blob/3eebe0542439384d29c4e9e2e1e1c2e7482533a9/pkg/ccl/streamingccl/streamproducer/event_stream.go#L365)):*
> ```Go
> 			} else if !iter.UnsafeKey().Less(endKey) {
> 				break
> 			}
> ```

This can be broken out from the above if/else as a standalone check.



_Code quote:_
```Go
			} else if !iter.UnsafeKey().Less(endKey) {
				break
			}
```
___
*[`pkg/ccl/streamingccl/streamproducer/event_stream.go` line 374 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N2M8klw5dX7qg8af0GA:-N2M8klw5dX7qg8af0GB:bnk9fl4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3eebe0542439384d29c4e9e2e1e1c2e7482533a9/pkg/ccl/streamingccl/streamproducer/event_stream.go#L374)):*
> ```Go
> }
> 
> // Add a RangeFeedSSTable into current batch and return size of the bytes added.
> ```

returns number of bytes added?
___
*[`pkg/ccl/streamingccl/streamproducer/event_stream.go` line 376 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N2M94Y_AWgiKunpCtnO:-N2M94Y_AWgiKunpCtnP:b6vs1zi) ([raw file](https://github.com/cockroachdb/cockroach/blob/3eebe0542439384d29c4e9e2e1e1c2e7482533a9/pkg/ccl/streamingccl/streamproducer/event_stream.go#L376)):*
> ```Go
> // Add a RangeFeedSSTable into current batch and return size of the bytes added.
> func (s *eventStream) addSST(sst *roachpb.RangeFeedSSTable, batch *streampb.StreamEvent_Batch) (int, error) {
> 	for _, sp := range s.spec.Spans {
> ```

Probably okay since the assumption is that the number of spans is small; but that assumption
may not be correct -- we may be resuming from checkpoint; 
Consider storing spanGroup for quick lookups?
___
*[`pkg/kv/bulk/sst_batcher.go` line 179 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N2M9QE82dmhLIqRYYwb:-N2M9QE82dmhLIqRYYwc:b-vgt6yx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fdbb16fa206c5fd77c3aa111ec40916edfb55df/pkg/kv/bulk/sst_batcher.go#L179)):*
> ```Go
> // AddMVCCKey adds a key+timestamp/value pair to the batch (flushing if needed).
> // This is only for callers that want to control the timestamp on individual
> // keys -- like RESTORE where we want the restored data to look the like backup.
> ```

I think it's the change in the comment...
___
*[`pkg/ccl/streamingccl/streamproducer/replication_stream_test.go` line 422 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N0OyuM1H3qMUL4CGzTn-r2-422:-N2MA8RX3VWcL01M-63F:b-7nia0m) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6eeaca5de164264f0d951d7ceb4a359f67c0a52/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go#L422)):*
<details><summary><i>Previously, gh-casper wrote…</i></summary><blockquote>

I previously did an import for one line "42, 42" and set the "kv.bulk_io_write.small_write_size" to be one which makes an import of any size to be an AddSSTable operation.
</blockquote></details>

Would it be useful to do this setting here, in this test? Better yet, have a method "ForAddSSTableForAllImport" or something like this?
___
*[`pkg/ccl/streamingccl/streamproducer/replication_stream_test.go` line 453 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N2MAuFS2kbrWws2ro42:-N2MAuFS2kbrWws2ro43:bfyfxnw) ([raw file](https://github.com/cockroachdb/cockroach/blob/3eebe0542439384d29c4e9e2e1e1c2e7482533a9/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go#L453)):*
> ```Go
> 		testAddSSTable(t, false, /* initialScan*/
> 			false /* addSSTableBeforeRangefeed */, "x4" /* table */)
> 	})
> ```

```
for _, initialScan := range []bool{true, false} {
   for _, addBefore := range []bool{true, false} {
      name := fmt.Sprintf("initialScan=%t/addBefore=%t", intialScan, addBefore)
      t.Run(name, 	t.Run("initial-scan-import-after-rangefeed", func(t *testing.T) {
		testAddSSTable(t, initialScan, addBefore, name)
	})
    }
}
```



<!-- Sent from Reviewable.io -->

gh-casper reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80450)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @miretskiy)

___
*[`pkg/ccl/streamingccl/streamproducer/event_stream.go` line 258 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N2M7FSh5U2yR5a0Npyi:-N2TiaKS1UDhqYke1YiC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3eebe0542439384d29c4e9e2e1e1c2e7482533a9/pkg/ccl/streamingccl/streamproducer/event_stream.go#L258)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

you probably want to select to make sure you detect ctx.Done
</blockquote></details>

Done.
___
*[`pkg/ccl/streamingccl/streamproducer/event_stream.go` line 335 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N2M8fCV9h48bu-PfC9D:-N2TijfsCChhHmvBVAuG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3eebe0542439384d29c4e9e2e1e1c2e7482533a9/pkg/ccl/streamingccl/streamproducer/event_stream.go#L335)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

add a function comment please?
Also wrap long lines if you can.
</blockquote></details>

Done.
___
*[`pkg/ccl/streamingccl/streamproducer/event_stream.go` line 365 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N2M8Nre5yv_cegN4YDU:-N2TlGZ-0trcbAi81fzB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3eebe0542439384d29c4e9e2e1e1c2e7482533a9/pkg/ccl/streamingccl/streamproducer/event_stream.go#L365)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

This can be broken out from the above if/else as a standalone check.
</blockquote></details>

Done.
___
*[`pkg/ccl/streamingccl/streamproducer/event_stream.go` line 374 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N2M8klw5dX7qg8af0GA:-N2TimCY1168GoW9BvWU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3eebe0542439384d29c4e9e2e1e1c2e7482533a9/pkg/ccl/streamingccl/streamproducer/event_stream.go#L374)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

returns number of bytes added?
</blockquote></details>

Done.
___
*[`pkg/ccl/streamingccl/streamproducer/event_stream.go` line 376 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N2M94Y_AWgiKunpCtnO:-N2NbjoP5j5maRB0EpTX:b-agcs5y) ([raw file](https://github.com/cockroachdb/cockroach/blob/3eebe0542439384d29c4e9e2e1e1c2e7482533a9/pkg/ccl/streamingccl/streamproducer/event_stream.go#L376)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Probably okay since the assumption is that the number of spans is small; but that assumption
may not be correct -- we may be resuming from checkpoint; 
Consider storing spanGroup for quick lookups?
</blockquote></details>

SpanGroup`Slice()` uses for loop to generate a new slice so I introduced `ForEach()` to support iterating it. 
___
*[`pkg/kv/bulk/sst_batcher.go` line 179 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N2M9QE82dmhLIqRYYwb:-N2TisLH4b6f93EMqsXg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fdbb16fa206c5fd77c3aa111ec40916edfb55df/pkg/kv/bulk/sst_batcher.go#L179)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I think it's the change in the comment...
</blockquote></details>

Done.
___
*[`pkg/ccl/streamingccl/streamproducer/replication_stream_test.go` line 422 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N0OyuM1H3qMUL4CGzTn-r2-422:-N2TlBA5EchZ5K6dhLLv:b-hbwoic) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6eeaca5de164264f0d951d7ceb4a359f67c0a52/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go#L422)):*
> !iter.UnsafeKey().Less(endKey) {
Added in the `testAddSSTable`.
___
*[`pkg/ccl/streamingccl/streamproducer/replication_stream_test.go` line 453 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N2MAuFS2kbrWws2ro42:-N2TjM_x1iZwIc3w0X4D:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3eebe0542439384d29c4e9e2e1e1c2e7482533a9/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go#L453)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

```
for _, initialScan := range []bool{true, false} {
   for _, addBefore := range []bool{true, false} {
      name := fmt.Sprintf("initialScan=%t/addBefore=%t", intialScan, addBefore)
      t.Run(name, 	t.Run("initial-scan-import-after-rangefeed", func(t *testing.T) {
		testAddSSTable(t, initialScan, addBefore, name)
	})
    }
}
```

</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (approved):
> Reviewed 1 of 11 files at r2, 4 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80450)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @gh-casper, and @miretskiy)

___
*[`pkg/ccl/streamingccl/event.go` line 135 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N2WHxRc5K8RuRjDSU-i:-N2WHxRc5K8RuRjDSU-j:bhrrkhn) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc4f3d47c0c4e007bf232be1055f9c7ce4c5d2af/pkg/ccl/streamingccl/event.go#L135)):*
> ```Go
> 
> // generationEvent indicates that the topology of the stream has changed.
> type generationEvent struct{}
> ```

We no longer have those events, do we? This needs to be cleaned up - another pr.
___
*[`pkg/ccl/streamingccl/streamproducer/event_stream.go` line 382 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N2WIX7Q1vp8G7_nEdS3:-N2WIX7Q1vp8G7_nEdS4:b-tfcdk7) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc4f3d47c0c4e007bf232be1055f9c7ce4c5d2af/pkg/ccl/streamingccl/streamproducer/event_stream.go#L382)):*
> ```Go
> func (s *eventStream) addSST(sst *roachpb.RangeFeedSSTable, batch *streampb.StreamEvent_Batch) (int, error) {
> 	// We send over the whole SSTable if the sst span is within
> 	// the subscribed spans boundary
> ```

nit "." at the end.
___
*[`pkg/ccl/streamingccl/streamproducer/event_stream.go` line 390 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80450#-N2WIeoZ-lcFzI3Z8G_R:-N2WIeoZ-lcFzI3Z8G_S:brkhmz4) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc4f3d47c0c4e007bf232be1055f9c7ce4c5d2af/pkg/ccl/streamingccl/streamproducer/event_stream.go#L390)):*
> ```Go
> 	// we trim the sst data to avoid sending unnecessary data.
> 	size := 0
> 	if err := s.trimSST(sst, func(keyVal roachpb.KeyValue) {
> ```

Drop a TODO to add metrics.  We want to track number of SSTs, and number of ssts that are not inside the boundaries (and possible count+size of kvs in such ssts).


<!-- Sent from Reviewable.io -->

gh-casper commented:
bors r+
